### ISSUE TITLE
Come up with an idea for generic mutable references

### ISSUE THREAD

**Issue by Robbepop** [Robbepop/pdsl#1]
# Generic Mutable References

## Problem
Currently wrappers around objects stored on the contract storage do not provide a way to receive mutable references (`&mut T`) to those objects.

## Rational

The reason for this is that users could do arbitrary changes to the objects without having an automated way to synchronize the storage.

## Example

Storage wrappers like `storage::Vec` and `storage::HashMap` provide APIs like `mutate_with` with which it is possible for users to do an efficient in-place mutation of an object owned by the collection.

## Goal

This is the tracking issue to come up with a user friendly, efficient and possibly general way to allow for mutable references that keep their referenced objects in sync between memory and storage.

### Solution 1: Do not provide mutable reference wrappers

This is the naive approach that we currently use.
However, it has the downside that it simply doesn't support what some users would like to have.

**Comment by Robbepop** [comment]
Today I tried out a solution that makes use of two new traits

```rust
/// Types that are mutable by some closure.
pub trait Mutate<T> {
	fn mutate<F>(&mut self, f: F)
	where
		F: FnOnce(&mut T);
}

/// Types that are assignable to some new value.
pub trait Assign<T> {
	fn assign(&mut self, val: T);
}
```

With these we could define different ref-mut types for all the different data structures, such as `storage::{Vec, HashMap, Stash, Value, ..}` and wrap them with another unifying and general `storage::RefMut<T>` that implements many `core` traits depending on whether `T` implements them so we could theoretically write something like this:

```rust
let mut v = my_storage_vec(); // storage::Vec<i32>
v.get_mut(5).unwrap() += 42;
```

The downside to this is that due to the nature of `core::ops::IndexMut` we cannot write something like this:

```rust
let mut v = my_storage_vec(); // storage::Vec<i32>
v[5] += 42;
```

Note that this implementation of a mutable reference to storage items works without having a proper flushing system.
With a proper flushing system that flushes all synchronized state to the contract storage as soon as there is a context switch (i.e. due to calling a remote contract) we could simply use `&mut T` as our mutable references. Wrapping of them and forwarding state synchronization would no longer be necessary since we would flush their state to contract storage anyway as soon as needed.

Although the approach implemented today is already pretty nice in certain ways I think we really want a proper flushing system for the best usability we could achieve.

**Comment by Vurich** [comment]
Why not just store the state inside `env` and make it so that `call_remote` is a method on `env` requiring `&mut`? That way you'd get a notification at compile-time that you have overlapping mutable references.

**Comment by Vurich** [comment]
To come back to this, the real root problem is that the ownership graph is all over the place. `call_remote` (essentially) needs mutable access to the state in the general case, if it was inside the same process this would be cut and dry - you either pass in a mutable reference to `call_remote` or you make `call_remote` an `&mut self` method on a struct that also owns the state. The only reason we're considering other options is because we're treating this like we're making an HTTP request, which is not how this works conceptually (HTTP requests are asynchronous and so don't have this issue, you can lock the state without causing a deadlock as long as unlocking the state doesn't wait on the request returning). Since `call_remote` is synchronous we have to treat it as if it were an in-process function call, from an API design perspective.

**Comment by Robbepop** [comment]
This is very valuable feedback!
Having compile time checks would be greatly appreciated we should really try to target having them.
For this to work we need to change how `Env` in `pdsl` interoperates with storage allocation.
For example, at the moment `Key` is seen as `raw pointer` having no lifetime that could be bound to a certain environment which was a sane default since environments so far were (stateless) wrappers around the SRML provided interfaces. However, if we want to make this change (and I really want to experiment with this) we would need to change that.

**Comment by Robbepop** [comment]
Btw.: The flushing system has already been implemented. I am currently trying to resolve some issues around storage allocation and initialization and will then continue with this.

**Comment by Robbepop** [comment]
Note that since the flushing system has been successfully implemented and with commit https://github.com/Robbepop/pdsl/commit/c3a942be1cedcb3544362be8a96b9e4ad07e8fa7 even `storage::Value` now has a `DerefMut` implementation which allows for very abstract mutable accesses just like in normal Rust code.

So far it seems to work very well.

tldr; I think we can close this issue.
